interface/patient_file/manage_dup_patients.php

```diff
@@ -262,21 +262,39 @@ function selchange(sel, toppid, rowpid) {
         array($_POST['form_toppid'])
     );
 } else if ($form_action == 'R') {
-    updateDupScore($_POST['form_toppid']);
+    updateDupScore($_POST['form_toppid']); /* this is not a new patient so check against all other patients */
 }
 
+/* generate a new table of possible duplicate patients */
 $query = "SELECT * FROM patient_data WHERE dupscore > 7 " .
     "ORDER BY dupscore DESC, pid DESC LIMIT 100";
 $res1 = sqlStatement($query);
+$used = array(); /* record which patients we display in the table, so as not to use them more than once */
 while ($row1 = sqlFetchArray($res1)) {
-    displayRow($row1);
-    $query = "SELECT p2.*, ($scorecalc) AS myscore " .
-    "FROM patient_data AS p1, patient_data AS p2 WHERE " .
-    "p1.pid = ? AND p2.pid < p1.pid AND ($scorecalc) > 7 " .
-    "ORDER BY myscore DESC, p2.pid DESC";
-    $res2 = sqlStatement($query, array($row1['pid']));
-    while ($row2 = sqlFetchArray($res2)) {
-        displayRow($row2, $row1['pid']);
+    global $used;
+    error_log ("p1 & score ".$row1['pid']." , ". $row1['dupscore']);
+    if (array_search($row1['pid'],$used) === false){ //haven't used this one yet
+    // build the sql query to compare to all other patients
+        $query = "SELECT p2.*, ($scorecalc) AS myscore " .
+        "FROM patient_data AS p1, patient_data AS p2 WHERE " .
+  /* check against all other patients - e.g. when demogrphics changed */
+        "p1.pid = ? AND p2.pid != p1.pid AND ($scorecalc) > 7 " .
+        "ORDER BY myscore DESC, p2.pid DESC";
+        $res2 = sqlStatement($query, array($row1['pid']));
+    /*only if more patients are potential duplicates display the first one -correct bug where only 1 patient in agroup displayed*/
+        $displayFirst = true;
+        while ($row2 = sqlFetchArray($res2)) {
+            error_log ("p2 & score ".$row2['pid']." , ".$row2['dupscore']);
+            if ($row2['dupscore'] != -1) { // don't use if marked as unique
+                if ($displayFirst )
+                {
+                    displayRow($row1);
+                    $displayFirst = false;
+                }
+                displayRow($row2, $row1['pid']);
+                $used[] = $row2['pid'];
```

Comment:
if you do the following $used['pid'] = $row2['pid'] you can remove the array search on line 276 and just change it to be the following:
```
if (!empty($used[$row2['pid']])) {
}
```
This will be more performant since you're going to be working with a bunch of patients.

Reply:
thanks that's useful - yes we do have quite large data sets!